[quant][fx][improvement] Add lowering support for BatchNormQuantizeHandler

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72490
* #72489

Summary:
This is an effort to move the current implementation towards the reference quantized model design:
https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md
so that we use reference model in the default fbgemm/qnnpack path

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps.test_qbatch_norm

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34062365](https://our.internmc.facebook.com/intern/diff/D34062365)

jerryzh168 committed:
[quant][fx][improvement] Add lowering support for BatchNormQuantizeHandler

Summary:
This is an effort to move the current implementation towards the reference quantized model design:
https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md
so that we use reference model in the default fbgemm/qnnpack path

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps.test_qbatch_norm

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34062365).

jerryzh168 committed:
Update on "[quant][fx][improvement] Add lowering support for BatchNormQuantizeHandler"


Summary:
This is an effort to move the current implementation towards the reference quantized model design:
https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md
so that we use reference model in the default fbgemm/qnnpack path

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps.test_qbatch_norm

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34062365](https://our.internmc.facebook.com/intern/diff/D34062365)

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -41,13 +41,21 @@ def is_fixed_qparams_node(node, modules):
 
+# TODO: merge with LOWER_MODULE_MAP after we merge
+# _lower_weighted_ref_module and special_pattern_replacement
+LOWER_MODULE_MAP2 = {
```
maybe name it `LOWER_MODULE_MAP_TEMP`?

andrewor14 replied:
or maybe `SPECIAL_PATTERN_LOWER_MODULE_MAP` if we're using it for `special_pattern_replacement` only

vkuzo commented on torch/nn/quantized/modules/batchnorm.py:
```diff
@@ -26,6 +26,19 @@ def from_float(cls, mod):
+            dtype=bn.weight.dtype
+        )
+        qbn.scale = float(output_scale)
+        qbn.zero_point = float(output_zero_point)
```
should this be int?

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -41,13 +41,21 @@ def is_fixed_qparams_node(node, modules):
 # Mapping from reference module class to the replacement quantized module class for lowering
-LOWER_MODULE_MAP: Dict[Type[nn.Module], Type[ReferenceableQuantizedModule]] = {
+# TODO: fix typing, the key is reference module
+LOWER_MODULE_MAP: Dict[Type[torch.nn.Module], Type[ReferenceableQuantizedModule]] = {
```
I see, yeah I'm not sure why we need this type, I thought it's supposed to be referencemodule, should we remove `ReferenceableQuantizedModule`?

andrewor14 replied:
Well the problem is we need to call `from_reference` in `_lower_weighted_ref_module`, but a `torch.nn.Module` doesn't have this method, so I feel we do need a common abstract class to declare this method

andrewor14 replied:
(Either that or we check `hasattr(module, "from_reference")`, but I feel the abstract class is cleaner)

jerryzh168 committed:
Update on "[quant][fx][improvement] Add lowering support for BatchNormQuantizeHandler"


Summary:
This is an effort to move the current implementation towards the reference quantized model design:
https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md
so that we use reference model in the default fbgemm/qnnpack path

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps.test_qbatch_norm

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34062365](https://our.internmc.facebook.com/intern/diff/D34062365)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][improvement] Add lowering support for BatchNormQuantizeHandler"


Summary:
This is an effort to move the current implementation towards the reference quantized model design:
https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md
so that we use reference model in the default fbgemm/qnnpack path

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps.test_qbatch_norm

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34062365](https://our.internmc.facebook.com/intern/diff/D34062365)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34062365).

jerryzh168 committed:
Update on "[quant][fx][improvement] Add lowering support for BatchNormQuantizeHandler"


Summary:
This is an effort to move the current implementation towards the reference quantized model design:
https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md
so that we use reference model in the default fbgemm/qnnpack path

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps.test_qbatch_norm

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34062365](https://our.internmc.facebook.com/intern/diff/D34062365)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][improvement] Add lowering support for BatchNormQuantizeHandler"


Summary:
This is an effort to move the current implementation towards the reference quantized model design:
https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md
so that we use reference model in the default fbgemm/qnnpack path

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps.test_qbatch_norm

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34062365](https://our.internmc.facebook.com/intern/diff/D34062365)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][improvement] Add lowering support for BatchNormQuantizeHandler"


Summary:
This is an effort to move the current implementation towards the reference quantized model design:
https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md
so that we use reference model in the default fbgemm/qnnpack path

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps.test_qbatch_norm

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34062365](https://our.internmc.facebook.com/intern/diff/D34062365)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][improvement] Add lowering support for BatchNormQuantizeHandler"


Summary:
This is an effort to move the current implementation towards the reference quantized model design:
https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md
so that we use reference model in the default fbgemm/qnnpack path

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps.test_qbatch_norm

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34062365](https://our.internmc.facebook.com/intern/diff/D34062365)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34062365).

jerryzh168 committed:
Update on "[quant][fx][improvement] Add lowering support for BatchNormQuantizeHandler"


Summary:
This is an effort to move the current implementation towards the reference quantized model design:
https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md
so that we use reference model in the default fbgemm/qnnpack path

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps.test_qbatch_norm

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34062365](https://our.internmc.facebook.com/intern/diff/D34062365)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34062365).

jerryzh168 committed:
Update on "[quant][fx][improvement] Add lowering support for BatchNormQuantizeHandler"


Summary:
This is an effort to move the current implementation towards the reference quantized model design:
https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md
so that we use reference model in the default fbgemm/qnnpack path

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps.test_qbatch_norm

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34062365](https://our.internmc.facebook.com/intern/diff/D34062365)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34062365).
